h2o-algos/src/test/java/hex/grid/GridSearchWithFilteringTest.java

```diff
@@ -0,0 +1,137 @@
+package hex.grid;
+
+import hex.grid.filter.KeepOnlyFirstMatchFilterFunction;
+import hex.grid.filter.PermutationFilterFunction;
+import hex.tree.gbm.GBMModel;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import water.TestUtil;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Note: exact type of {@code Model.Parameters} is transparent to filtering functionality 
+ * so we can test on any of the existing ones.
+ */
+public class GridSearchWithFilteringTest extends TestUtil {
+
+  @BeforeClass
+  public static void setup() {
+    stall_till_cloudsize(1);
+  }
+  
+  @Test
+  public void randomGridSearchWithFunctionsToFilterOutUnnecessaryGridItems_GBMParametersGrid() {
+    Map<String, Object[]> hpGrid = new HashMap<>();
+    hpGrid.put("_ntrees", new Integer[]{10000});
+    hpGrid.put("_max_depth", new Integer[]{3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17}); 
+    hpGrid.put("_min_rows", new Integer[]{1, 5, 10, 15, 30, 100}); 
+    hpGrid.put("_learn_rate", new Double[]{0.001, 0.005, 0.008, 0.01, 0.05, 0.08, 0.1, 0.5, 0.8}); 
+    hpGrid.put("_sample_rate", new Double[]{0.50, 0.60, 0.70, 0.80, 0.90, 1.00}); 
+    hpGrid.put("_col_sample_rate", new Double[]{ 0.4, 0.7, 1.0}); 
+    hpGrid.put("_col_sample_rate_per_tree", new Double[]{ 0.4, 0.7, 1.0}); 
+    hpGrid.put("_min_split_improvement", new Double[]{1e-4, 1e-5});
+
+    int totalNumberOfPermutationsInHPGrid = 87480;
+
+    GBMModel.GBMParameters gbmParameters = new GBMModel.GBMParameters();
+
+    GridSearch.SimpleParametersBuilderFactory simpleParametersBuilderFactory = new GridSearch.SimpleParametersBuilderFactory();
+
+    HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria hyperSpaceSearchCriteria = new HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria();
+
+    PermutationFilterFunction strictFilterFunction = gridItem -> {
+      return !((int) gridItem.get("_max_depth") >= 15.0 && (int) gridItem.get("_min_rows") >= 50);
+    };
+    int expectedNumberOfFilteredOutPermutations = 2916;
+
+    ArrayList<PermutationFilterFunction> filterFunctions = new ArrayList<>();
```

Comment:
even in tests it is good practice to declare variables with interfaces and use implementations only in instatiations

Reply:
repeats countless times in other files as well

Reply:
@honzasterba I agree that it is a best practice, but of a less importance in the test. It is more important with signatures and returned values of the functions, APIs. Try to search for `new ArrayList<>()` in our code base and you will see that the majority of the matches don't follow it ( no matter whether it is a test or source code). Will fix.

Reply:
Yes, most of our code is poorly written, please don't add more bad code.

Reply:
@honzasterba please, no need to be rude in reviews. Just stick to the code.
`it is good practice to declare variables with interfaces and use implementations only in instantiations`
True, and no need to say more.
@deil87 please, no need to justify everything: `I agree that it is a best practice [...]. Will fix`. That's all we want to know, thanks.

Reply:
@seb-h2o I am sorry if you find my comment rude, I definitely did not mean to offend anyone. You are right, there was no need to say anything. I just got carried away by the endless discussions.